lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1595,6 +1649,169 @@ class BuildBBRegs : public FuncOpPartialLoweringPattern {
   }
 };
 
+/// Builds registers for each pipeline stage in the program.
+class BuildPipelineRegs : public FuncOpPartialLoweringPattern {
+  using FuncOpPartialLoweringPattern::FuncOpPartialLoweringPattern;
+
+  LogicalResult
+  PartiallyLowerFuncToComp(mlir::FuncOp funcOp,
+                           PatternRewriter &rewriter) const override {
+    funcOp.walk([&](staticlogic::PipelineRegisterOp op) {
+      // Condition registers are handled in BuildWhileGroups.
+      auto *parent = op->getParentOp();
+      auto stage = dyn_cast<staticlogic::PipelineStageOp>(parent);
+      if (!stage)
+        return;
+
+      // Create a register for each stage.
+      for (size_t i = 0; i < op.getNumOperands(); ++i) {
+        // Iter args are created in BuildWhileGroups, so just mark the iter arg
+        // register as the appropriate pipeline register.
+        Value stageResult = stage.getResult(i);
+        bool isIterArg = false;
+        for (auto &use : stageResult.getUses()) {
+          if (auto term =
+                  dyn_cast<staticlogic::PipelineTerminatorOp>(use.getOwner())) {
+            if (use.getOperandNumber() < term.iter_args().size()) {
```

Comment:
This seems a bit dangerous; can an assert be added to sanity check here?

Reply:
I'm not sure what we should assert in this case. The terminator has operands for both iter args and results, and this is checking that we are considering an iter arg.